internal/luks2/metadata.go

```diff
@@ -0,0 +1,874 @@
+		}
+		m.Keyslots[id] = v
+	}
+
+	m.Segments = make(map[int]*Segment)
+	for k, v := range d.Segments {
+		id, err := k.int()
+		if err != nil {
+			return xerrors.Errorf("invalid segment index: %w", err)
+		}
+		m.Segments[id] = v
+	}
+
+	m.Digests = make(map[int]*Digest)
+	for k, v := range d.Digests {
+		id, err := k.int()
+		if err != nil {
+			return xerrors.Errorf("invalid digest index: %w", err)
+		}
+		m.Digests[id] = v
+	}
+
+	m.Tokens = make(map[int]*Token)
+	for k, v := range d.Tokens {
+		id, err := k.int()
+		if err != nil {
+			return xerrors.Errorf("invalid token index: %w", err)
+		}
+		m.Tokens[id] = v
+	}
+
+	m.Config = d.Config
+	return nil
+}
+
+// HdrInfo corresponds to the header (binary header and JSON metadata) for a LUKS2 volume.
+type HdrInfo struct {
+	HdrSize  uint64   // The total size of the binary header and JSON metadata in bytes
+	Label    string   // The label
+	Metadata Metadata // JSON metadata
+}
+
+func decodeAndCheckHeader(r io.ReadSeeker, offset int64, primary bool) (*binaryHdr, []byte, error) {
+	if _, err := r.Seek(offset, io.SeekStart); err != nil {
+		return nil, nil, err
+	}
+
+	var hdr binaryHdr
+	if err := binary.Read(r, binary.BigEndian, &hdr); err != nil {
+		return nil, nil, xerrors.Errorf("cannot read header: %w", err)
+	}
+	switch {
+	case primary && bytes.Equal(hdr.Magic[:], []byte("LUKS\xba\xbe")):
+	case !primary && bytes.Equal(hdr.Magic[:], []byte("SKUL\xba\xbe")):
+	default:
+		return nil, nil, errors.New("invalid magic")
+	}
+	if hdr.Version != 2 {
+		return nil, nil, errors.New("invalid version")
+	}
+	if hdr.HdrSize > uint64(math.MaxInt64) {
+		return nil, nil, errors.New("header size too large")
+	}
+	if hdr.HdrOffset > uint64(math.MaxInt64) {
+		return nil, nil, errors.New("header offset too large")
+	}
+	if int64(hdr.HdrOffset) != offset {
+		return nil, nil, errors.New("invalid header offset")
+	}
+
+	// Verify the header checksum, which includes the JSON metadata
+	csumHash := hdr.CsumAlg.GetHash()
+	if csumHash == 0 {
+		return nil, nil, errors.New("unsupported checksum alg")
+	}
+
+	h := csumHash.New()
+
+	// Hash the binary header without the checksum
+	hdrTmp := hdr
+	hdrTmp.Csum = [64]byte{}
+	if err := binary.Write(h, binary.BigEndian, &hdrTmp); err != nil {
+		return nil, nil, xerrors.Errorf("cannot calculate checksum, error serializing header: %w", err)
+	}
+
+	// Hash the JSON metadata area, keeping a copy of the hashed metadata in memory
+	var jsonBuffer bytes.Buffer
+	tr := io.TeeReader(r, &jsonBuffer)
+	if _, err := io.CopyN(h, tr, int64(hdr.HdrSize)-int64(binary.Size(hdr))); err != nil {
+		return nil, nil, xerrors.Errorf("cannot calculate checksum, error reading JSON metadata: %w", err)
+	}
+
+	if !bytes.Equal(h.Sum(nil), hdr.Csum[0:csumHash.Size()]) {
+		return nil, nil, errors.New("invalid header checksum")
+	}
+
+	// Return the binary header and the memory buffer containing the verified JSON metadata
+	return &hdr, jsonBuffer.Bytes(), nil
```

Comment:
from the usage it seems this could return the jsonBuffer directly instead of Bytes of it?

Reply:
I've updated the code to do this now.